modules/core/upgrade.py

```diff
@@ -0,0 +1,551 @@
+        """.format(
+            user=username,
+            addr=ip_addr,
+            lcertpath=local_ca_cert_path,
+            lcertname=os.path.basename(local_ca_cert_path),
+            rcertdir=remote_ca_cert_dir))
+    except:
+        raise AssertionError(
+            'Unable to send CA certificate across to "{}"!'.format(ip_addr))
+
+
+def unseal_vault(username, ip_addr, vault_local_addr, unseal_key):
+    try:
+        return invoke_shell("""
+        ssh -o StrictHostKeyChecking=no {user}@{addr} \
+            "vault operator unseal -address {local_addr} {unseal_key}"
+        """.format(
+            user=username,
+            addr=ip_addr,
+            local_addr=vault_local_addr,
+            unseal_key=unseal_key))
+    except:
+        raise AssertionError(
+            'Unable to unseal vault in "{}@{}", using Vault local address "{}" with given unseal key!'
+            .format(username, ip_addr, vault_local_addr))
+
+#
+# High level
+#
+
+
+def upgrade_consul(consul_tag_pattern, address, check_interval, timeout):
+    print("Upgrading Consul instances...")
+
+    # Sanity check
+    aws_instances = get_instance_ids_from_tag(consul_tag_pattern)
+    consul_nodes = list_consul_peers(address)
+    print('AWS instances: {}'.format(aws_instances))
+    print(' Consul nodes: {}'.format(consul_nodes))
+    assert_same_instances(aws_instances, consul_nodes)
+
+    n = find_n_to_kill_in_quorum(aws_instances)
+    assert_n_quorum(n)
+
+    def check_fn(prev_aws_instances, idx):
+        return check_service_up(prev_aws_instances, idx, n,
+                                consul_tag_pattern, lambda: list_consul_peers(address))
+
+    def post_fn(new_instances):
+        # Do nothing
+        pass
+
+    kill_check_post(
+        kill_fn,
+        check_fn,
+        post_fn,
+        consul_tag_pattern,
+        n,
+        check_interval,
+        timeout)
+
+
+def upgrade_nomad_server(nomad_server_tag_pattern, address, check_interval, timeout):
+    print("Upgrading Nomad Servers instances...")
+
+    # Sanity check
+    aws_instances = get_instance_ids_from_tag(nomad_server_tag_pattern)
+    nomad_servers = list_nomad_server_members(address)
+    print('AWS instances: {}'.format(aws_instances))
+    print('Nomad servers: {}'.format(nomad_servers))
+    assert_same_instances(aws_instances, nomad_servers)
+
+    n = find_n_to_kill_in_quorum(aws_instances)
+    assert_n_quorum(n)
+
+    def check_fn(prev_aws_instances, idx):
+        return check_service_up(prev_aws_instances, idx, n, nomad_server_tag_pattern,
+                                lambda: list_nomad_server_members(address))
+
+    def post_fn(new_instances):
+        # Do nothing
+        pass
+
+    kill_check_post(
+        kill_fn,
+        check_fn,
+        post_fn,
+        nomad_server_tag_pattern,
+        n,
+        check_interval,
+        timeout)
+
+
+# TODO - Need to figure out how to properly wait for all upgraded allocs get
```

Comment:
Instead of this, you can just check that the node going to be killed has no more running allocs.